### PR TITLE
Make blank target imply noopener, implement "opener" link type

### LINKED ISSUE #22845: Make target=_blank imply noopener; support opener
whatwg/html#4330
web-platform-tests/wpt#15188

### PR DESCRIPTION
**Pull request by georgeroman**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix part of #22845 

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22851)

### CODE REVIEW

File: components/script/dom/htmlanchorelement.rs [hunk 1]
```diff
@@ -609,7 +609,17 @@ pub fn follow_hyperlink(subject: &Element, hyperlink_suffix: Option<String>) {
         let values = link_types.Value();
         let contains_noopener = values.contains("noopener");
         let contains_noreferrer = values.contains("noreferrer");
-        contains_noreferrer || contains_noopener
+        let contains_opener = values.contains("opener");
+        let target_is_blank = if let Some(name) = target_attribute_value.as_ref() {
+            match name.Value().to_lowercase().as_ref() {
```
Comment (CYBAI) [inline components/script/dom/htmlanchorelement.rs]: nit: How about doing like following?

```rust
let target_is_blank = if let Some(name) = target_attribute_value.as_ref() {
    name.Value().to_lowercase() == "_blank"
} else {
  false
};
```

**Review by CYBAI — APPROVED** [review]
Other than the nit, the `Maketarget=_blank imply noopener` part looks fine to me. But, it seems we didn't support [rel=opener](https://html.spec.whatwg.org/multipage/#link-type-opener) yet so maybe let's update the PR description to `These changes fix part of #22845` ?

Besides, the WPT test PR for the spec update has not been merged yet. Let's wait the PR merged and then try wpt for this PR?

cc @jdm @gterzian

**Comment by gterzian** [comment]
> But, it seems we didn't support [rel=opener](https://html.spec.whatwg.org/multipage/#link-type-opener) yet so maybe let's update the PR description to `These changes fix part of #22845` ?

The way I understand the change of the spec, is the an explicit `rel=opener` is now required for a `target=_blank` to open an new auxiliary, wherease previously the `rel=opener` was implied, unless an explicit `rel=noopener` was present(and we supported this via https://github.com/servo/servo/blob/71ba247942a95d1511c9fa62f06d860fcc03de1c/components/script/dom/windowproxy.rs#L456.)

**Review by gterzian — CHANGES_REQUESTED** [review]
Thanks for the contribution! Looks good so far!

As suggested earlier, we should also wait for https://github.com/web-platform-tests/wpt/pull/15188 to merge, and for servo to sync this change(which will look like https://github.com/servo/servo/pull/22859)...

File: components/script/dom/htmlanchorelement.rs [hunk 2]
```diff
@@ -609,7 +609,14 @@ pub fn follow_hyperlink(subject: &Element, hyperlink_suffix: Option<String>) {
         let values = link_types.Value();
         let contains_noopener = values.contains("noopener");
         let contains_noreferrer = values.contains("noreferrer");
-        contains_noreferrer || contains_noopener
+        let contains_opener = values.contains("opener");
+        let target_is_blank = if let Some(name) = target_attribute_value.as_ref() {
```
[the remark this answers is not in the document]
Comment (gterzian) [reply components/script/dom/htmlanchorelement.rs]: It is indeed relevant, just seeing step 6 of https://html.spec.whatwg.org/multipage/links.html#following-hyperlinks-2

**Comment by gterzian** [comment]
I changed the title because I think this indeed implements the "opener" link type, since that is a result of checking for the presence of `opener` on a `_blank` link. Any link with a named target other than `_blank` will also have an `opener`(since the logic only sets `noopener` to `true` when the `opener` keyword is absent, **and it is a `_blank` target**...

**Comment by CYBAI** [comment]
@bors-servo try=wpt

- The tests has been merged at #22867. Let's try it!

**Comment by CYBAI** [comment]
Hmm, @georgeroman, could you rebase to latest master and let's try again? thanks.

**Comment by jdm** [comment]
Rebasing does not make a difference; try runs are merged with master, so the results reflect the most up to date code changes.

**Comment by pradyunsg** [comment]
Is this good to go?

**Comment by jdm** [comment]
I think we still would like to figure out why the tests aren't showing any different results.

**Comment by gterzian** [comment]
The test that should go from `FAIL` to `PASS` is at https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/html/semantics/links/links-created-by-a-and-area-elements/target_blank_implicit_noopener.html, and can be run with `./mach test-wpt /html/semantics/links/links-created-by-a-and-area-elements/target_blank_implicit_noopener.html`. 

The test expectation, that should be update if we get the test to pass, is at https://github.com/servo/servo/blob/master/tests/wpt/metadata/html/semantics/links/links-created-by-a-and-area-elements/target_blank_implicit_noopener.html.ini

**Comment by georgeroman** [comment]
I ran the tests and they all fail with the message: `BroadcastChannel is not defined`.

**Comment by jdm** [comment]
@bors-servo r=cybai,jdm
Thanks for checking!

**Pull request merged**